src/AzureAuth/AuthFlowResultExtensions.cs

```diff
@@ -0,0 +1,89 @@
+        /// without bringing a reference to the CLI framework into the <see cref="MSALWrapper"/>. This is not a boundary we want to cross.
+        /// Instead, here in AzureAuth, we can define the conversion using an extension method to house it rather than an arbitrary
+        /// static class to hold the same function. Being an extension here is purely syntactic sugar. The main preference is
+        /// around defining a "pure" function for this.
+        /// </remarks>
+        public static EventData EventData(this AuthFlowResult result)
+        {
+            if (result == null)
+            {
+                return null;
+            }
+
+            var eventData = new EventData();
+            eventData.Add("authflow", result.AuthFlowName);
+            eventData.Add("success", result.Success);
+            eventData.Add("duration_milliseconds", (int)result.Duration.TotalMilliseconds);
+
+            if (result.Errors.Any())
+            {
+                var error_messages = ExceptionListToStringConverter.Execute(result.Errors);
+                eventData.Add("error_messages", error_messages);
+            }
+
+            if (result.Success)
+            {
+                eventData.Add("msal_correlation_id", result.TokenResult.CorrelationID.ToString());
+                eventData.Add("token_validity_minutes", result.TokenResult.ValidFor.TotalMinutes);
+                eventData.Add("silent", result.TokenResult.IsSilent);
+            }
+
+            return eventData;
+        }
+
+        /// <summary>
+        /// Send telemetry events for n <see cref="AuthFlowResult"/> instance.
+        /// </summary>
+        /// <param name="results"><see cref="AuthFlowResult"/>s to send telemetry for.</param>
+        /// <param name="telemetryService">The <see cref="ITelemetryService"/> to use to send the events.</param>
+        /// /// <remarks>
+        /// Why is this an extension method? <see cref="AuthFlowResult"/> is owned by <see cref="MSALWrapper"/>
+        /// and <see cref="ITelemetryService"/> is owned by <see cref="Office.Lasso"/>.
+        /// So <see cref="AuthFlowResult"/> does not actually know what a <see cref="ITelemetryService"/> is
+        /// without bringing a reference to the CLI framework into the <see cref="MSALWrapper"/>. This is not a boundary we want to cross.
+        /// Instead, here in AzureAuth, we can define the conversion using an extension method to house it rather than an arbitrary
+        /// static class to hold the same function. Being an extension here is purely syntactic sugar. The main preference is
+        /// around defining a "pure" function for this.
+        /// </remarks>
+        public static void SendTelemetry(this IEnumerable<AuthFlowResult> results, ITelemetryService telemetryService)
+        {
+            foreach (var result in results)
+            {
+                telemetryService.SendEvent($"authflow_{result.AuthFlowName}", result.EventData());

```

Comment:
Is there a reason we reverted it back to not using `Parallel.ForEach`?

Reply:
Yes, good question: 2 reasons:
1. The telemetry service is not explicitly thread safe, so this was never truly safe but it seems it was working.
2. We're about to update Lasso with the change to buffer events and only start 1 process, and that is also not thread safe, but also does not benefit from concurrency, because we would have to use a Concurrent Buffer, and the adds would be synchronized. The intention here was that each call is spawning a process, and so there was a benefit to doing this in paralle.